api/v1/project_reference_cel_test.go

```diff
@@ -1,130 +1,174 @@
 package v1
 
 import (
+	"fmt"
+	"path/filepath"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/api"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/api/v1/common"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/test/helper/cel"
 )
 
-type CELTestObjectGenFunc func(*ProjectDualReference) AtlasCustomResource
+const (
+	crdBasePath = "../../config/crd/bases/"
+)
 
-type celValidationTestCase struct {
-	title          string
-	obj, old       AtlasCustomResource
-	expectedErrors []string
+type projectReferrer interface {
+	ProjectDualRef() *ProjectDualReference
 }
 
-func launchProjectRefCELTests(t *testing.T, genObjFunc CELTestObjectGenFunc, crdPath string) {
-	testCases := testCasesFor(genObjFunc)
-	for _, tc := range testCases {
-		t.Run(tc.title, func(t *testing.T) {
-			unstructuredObject, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&tc.obj)
-			require.NoError(t, err)
-
-			unstructuredOldObject, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&tc.old)
-			require.NoError(t, err)
-
-			validator, err := cel.VersionValidatorFromFile(t, crdPath, "v1")
-			assert.NoError(t, err)
-			errs := validator(unstructuredObject, unstructuredOldObject)
-
-			require.Equal(t, len(tc.expectedErrors), len(errs))
+type projectReferrerObject interface {
+	client.Object
```

Comment:
nit: `runtime.Object` is good enough as it satisfies `DeepCopyObject()`

Reply:
`client.Object` has accessors (getters/setters) to metadata which is not needed here.